### PR TITLE
Backport docs for System.Numerics.Vectors

### LINKED ISSUE #48968: System.Numerics: Backport MS Docs documentation to triple slash
We are working on a [new documentation process plan](https://github.com/dotnet/runtime/issues/44969#issuecomment-788536998), in which the main objective is to make triple slash comments the source of truth for documentation, instead of MS Docs: We want developers/maintainers to have an easier time maintaining the documentation for their APIs.

You can use the [DocsPortingTool](https://github.com/carlossanlop/DocsPortingTool) to automate the backport process: Run the tool targeting the assembly, then submit a PR with the changes.

You can find detailed instructions for the backporting process [here](https://github.com/carlossanlop/DocsPortingTool/blob/master/BackportInstructions.md).

Area owners are free to decide if they want to address this in 6.0 or in Future.

### PR DESCRIPTION
**Pull request by carlossanlop**
Backporting public API documentation from MS Docs into triple slash comments for the System.Numerics.Vectors namespace.

This namespace was interesting for this experiment because all the source files live in `System.Private.CoreLib`, so the porting tool had to make sure to read the `System.Numerics.Vectors.csproj` dependent projects and find the source code files in them.

The full description of the purpose of this PR is described in the issue [#44969 - Pilot a new process that extracts API documentation from source code](https://github.com/dotnet/runtime/issues/44969) under the section "Bring documentation from Docs to triple slash".

I ran the [porting tool](https://github.com/carlossanlop/DocsPortingTool) using this command:

```
DocsPortingTool.exe \
-CsProj D:\runtime\src\libraries\System.Numerics.Vectors\src\System.Numerics.Vectors.csproj \
-Docs D:\dotnet-api-docs\xml \
-Save true \
-Direction ToTripleSlash \
-IncludedAssemblies System.Numerics \
-IncludedNamespaces System.Numerics
```

Please compare the ported docs with their source files in dotnet-api-docs: 
https://github.com/dotnet/dotnet-api-docs/tree/master/xml/System.Numerics

### CODE REVIEW

**Comment by carlossanlop** [comment]
Need to refresh this branch. There are some recent nullability changes causing a conflict.

File: src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln
Comment (safern) [inline src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln:1]: I guess these changes to the sln file are not intentional?

Comment (carlossanlop) [reply src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln:1]: Visual Studio made those changes automatically (first commit in the PR).

File: src/libraries/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj
Comment (safern) [inline src/libraries/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj:6]: Is System.Private.CoreLib already generating a documentation file? We need it to generate it and disable warnings until we port all the documentation to System.Private.CoreLib.

**Comment by safern** [comment]
> This namespace was interesting for this experiment because all the source files live in System.Private.CoreLib, so the porting tool had to make sure to read the System.Numerics.Vectors.csproj dependent projects and find the source code files in them.

For implementation that lives in System.Private.CoreLib, we should run the tool with `RuntimeFlavor=Mono` so that specific API that is exposed via the Mono S.P.CoreLib if any, gets it's docs ported.

File: src/libraries/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj
Comment (carlossanlop) [reply src/libraries/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj:6]: I wanted to confirm this with you.

S.P.CoreLib does generate docs, but they are not yet made mandatory. I assume this change needs to be reverted, and the `GenerateDocumentationFile` change needs to be the last change we add to CoreLib, after all APIs that live there have their documentation backported.

**Comment by carlossanlop** [comment]
> we should run the tool with RuntimeFlavor=Mono

@safern are you saying I should execute the DocsPortingTool with this argument?

**Comment by safern** [comment]
> @safern are you saying I should execute the DocsPortingTool with this argument?

Yeah I'm saying the compilations when loading System.Private.CoreLib, should do a pass to get the locations with RuntimeFlavor=CoreClr (which is the default) and with RuntimeFlavor=Mono.

File: src/libraries/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj
Comment (safern) [reply src/libraries/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj]: I think having the `GenerateDocumentationFile=true` here is valuable as a future proof in the case where we add public API to this assembly and the implementation were to live here as opposed to System.Private.CoreLib.

We will have to add some infrastructure to merge the System.Private.CoreLib.xml file into the reference assemblies that expose those APIs xmls.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs
Comment (jeffhandley) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:200]: Typo: Should be "Multiplies"

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs
Comment (jeffhandley) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs:174]: Typo: Should be "Multiplies"

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs
Comment (jeffhandley) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs:185]: Typo: Should be "Multiplies"

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs
Comment (jeffhandley) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs:798]: There are a bunch of `&lt;` and `&gt;` contents below here. Should these be encoded differently?

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs
Comment (jeffhandley) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs:150]: Typo: Should be "Multiplies"

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs
Comment (jeffhandley) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs:161]: Typo: Should be "Multiplies"

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs
Comment (jeffhandley) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:211]: Typo: Should be "Multiplies"

**Review by jeffhandley — COMMENTED** [review]
There are several places with a typo of "Multiples" where it should be "Multiplies", as it was in the original `///` comments.

Should we fix the typos in `dotnet-api-docs`,  and re-run the doc port and update this PR? Or should we just fix those by-hand in this PR and wait for `dotnet-api-docs` to get rehydrated from the `///` comments after this is merged in?

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs:798]: Good catch. These should have been mentioned  as <`see cref="docid" />`.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:200]: Thanks for catching those. I think fixing them by hand is the best approach here. This wasn't a problem with the tool, but rather a real typo in the source.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs:798]: Note: This is also a case of something that came as is from MS Docs. I'll fix it manually here, so next time we sent these docs to MS Docs, they'll get the correct content.

Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs:798]: This particular case is referring to the type of the passed parameters, which is `Vector<byte>`. It seems I cannot mention them as a `see cref`, which is why these comments mention them this way.

I'll leave these specific cases unmodified, but I did find a few other similar cases in the same file that can be fixed.

File: src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln
Comment (pgovind) [reply src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln:1]: I think these changes occur when a newer version of VS opens the sln. AFAIK, reverting these changes work, so that's what I've always done.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs:798]: > It seems I cannot mention them as a see cref, which is why these comments were added this way.

You can, as `cref="Vector{Byte}" />`, the issue is that somewhere between Roslyn and msdocs processing the reference, this is reverted to simply `Vector{T}` and we lose the generic type information.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:135]: I don't think we should block, but I'm really not a fane of this style of summary. I think `Multiples` first, as the action the method is perform, reads significantly better

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:509]: Same here, I think `Attempts` or `Tries` first makes it more immediately obvious that this is
1. A badly named API that should have been called `TryInvert`
2. Is easier to read and ensures the user understand the operation/pattern

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:31]: In this case, I think there is probably a good in between between what we had and what is here.

Matrices are in terms of columns and rows and so I'd think "element" may be initially confusing to some users.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (pgovind) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:63]: What is this <value> tag? Is it new? Don't we use <return> here instead usually?

Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:63]: Properties need their value to be described. Here's the full explanation of this tag in the dotnet-api-docs Wiki:

https://github.com/dotnet/dotnet-api-docs/wiki/Property-Value

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:31]: As the area owner, feel free to provide a suggested text and I can apply it.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs:798]: Yep, I found that option to be ugly. I prefer to keep the generic type, so the `<c></c>` seems to be the most appropriate workaround.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:31]: I don't want to block this PR on it, since we aren't changing anything.

I think it might be good to see if we can get a best/recommended practices for documentation text though so we can try to:
1. Be overall consistent across .NET
2. Find something that works in cases where the terminology may end up differing

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Plane.cs
Comment (gewarren) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Plane.cs:23]:
```suggestion
        /// <summary>Creates a <see cref="T:System.Numerics.Plane" /> object from the X, Y, and Z components of its normal, and its distance from the origin on that normal.</summary>
```

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs
Comment (gewarren) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs:12]: Not sure about using \``` in XML. Also, it should have a code language.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs
Comment (gewarren) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs:154]:
```suggestion
        /// <returns>The resulting integral vector.</returns>
```

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:139]: Doesn't C# add these automatically? If not should it be updated to do so?

Comment (gewarren) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:139]: Oh yes, you are right. So this should be fine then. https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/cref-attribute

Comment (gewarren) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:139]: There's even a code analysis rule that enforces *not* setting a prefix - https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1200.

File: src/libraries/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj
Comment (carlossanlop) [reply src/libraries/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj]: Okay thanks. If you think it's safe to keep it, then I'll keep it.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:135]: @tannergooding if you can give me comment suggestions with the texts you prefer to see, that would be great. That way I can batch apply them.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:11]: @gewarren that's a good point. In the first PR, I was asked to remove the CDATA since it was not providing any value in triple slash comments, was using a lot of unnecessary space and polluting the code visually.

So I have a question for you to help address this: When we send intellisense xmls to the Docs build system, is it smart enough to automatically wrap non-empty remarks (including remarks that say `To be added.`) with these tags?:

 `<format type="text/markdown><![CDATA[ ... ]]></format>` 

If not, then we have two options:

A) Convert markdown links like these to valid xml. Do you know what's the non-markdown syntax for a link to an md file? And do you know if the Docs build system will be able to read that correctly and render the md document properly?
B) Talk to the Docs dev team to request converting any remarks we send to them to markdown (if they don't do it yet).

Thoughts?

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs:12]: Good catch. These are the cases that we need to manually move to a file in dotnet-api-docs, then link to it here.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:135]: I think either `Multiplies two matrices together to compute the product` is simple, explanatory, and puts it in common math terms.

Likewise, `Divides two matrices to compute the quotient`, `Adds two matrices to compute the sum`, `Subtracts two matrices to compute the difference`.

And when the types differ: `Multiplies a matrix by a float to compute the product` or `Divides a matrix by a float to compute the quotient`, etc

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:135]: (Assuming, of course, the docs team doesn't see any issues here)

File: src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln
Comment (carlossanlop) [reply src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln:1]: It's kind of annoying having to revert the changes every time I have to submit a new commit. @safern would I break something if I keep the sln changes?

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:11]: Another thing to consider is that the xmls that Roslyn will produce after the source of truth are comments in source code, we will be shipping that as part of the intellisense xml files right? How will this look when a customer hovers over an API on VS and VS takes the info for that API from comments with these includes or `[CDATA]`?

Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:11]: AFAIK, remarks do not show up in VS intellisense in any way.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (gewarren) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:135]: Tanner's suggestions sound great to me.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (gewarren) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:11]: > Do you know what's the non-markdown syntax for a link to an md file?

I don't know. And the thing is, you don't just want a link, you actually want to import the contents of the linked file.

> When we send intellisense xmls to the Docs build system, is it smart enough to automatically wrap non-empty remarks?

No, it does not. See for example [here](https://github.com/dotnet/dotnet-api-docs/blob/733b2cecdd45ab3ed5efee2f001d9b07730f06e1/xml/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckPublisherOptions.xml#L117).

Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:11]: @gewarren to address this problem, we can keep the CDATA section for remarks that require to show an include.

Comment (gewarren) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:11]: That's what the machinelearning repo does for its linked code snippets. See https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/TreeTrainersCatalog.cs#L30.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:139]: @gewarren There is a bug preventing me from indicating a method group in triple slash comments (the equivalent of `Overload:` as you mentioned). There is an open issue tracking this problem in the csharplang repo: https://github.com/dotnet/csharplang/issues/320

Would you mind helping me increase the priority/visibility of the csharplang issue? I am currently running the tool in other larger assemblies, and this is going to become a very common problem when porting types with lots of overloads (Process, FileStream are giving me a headache).

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs:12]: Found a bunch of these and I have converted them either to remarks with format + cdata, or to `<c></c>`.

**Comment by carlossanlop** [comment]
@gewarren @tannergooding @pgovind I marked this draft as ready for review. I addressed all your comments and suggestions and made sure the remarks with special markdown data show their information properly. Can you please take a look?

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs]: I addressed the ones I could find in Matrix3x2 and in Matrix4x4. If you find more, please help me by using the `` ```suggestion`` feature from GitHub.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:33]: Not sure what the action item is here. But if you'd like us to improve the docs, feel free to provide `` ```suggestion`` comments so I can bulk apply them.

File: src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln
Comment (carlossanlop) [reply src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln:1]: I'll keep this unless it causes any issues.

File: src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln
Comment (safern) [inline src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln:34]: I don't understand why this is being added? Did you change this on VS?

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs
Comment (gewarren) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:141]: Done!

**Review by tannergooding — APPROVED** [review]

File: src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln
Comment (carlossanlop) [reply src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln:34]: No, I didn't modify the sln file myself. Visual Studio did this. I'll revert the sln file change.

Comment (carlossanlop) [reply src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln]: @safern reverted.

**Comment by carlossanlop** [comment]
Learnings from this PR:

- Overloads can't be shown in triple slash comments. This is a bug, and it is tracked by https://github.com/dotnet/csharplang/issues/320 .
- When remarks contain a hyperlink or an include to an example, the whole element needs to be transformed to markdown (wrapped by `<format type="text/markdown"><![CDATA[ ... ]]></format>`).

**Comment by carlossanlop** [comment]
The two CI failures are unrelated to my change. I already tried re-running both legs 4 times and the failure is consistent.

One CI leg is failing with this:
```
ERROR: helix-scripts 20191204.1 has requirement azure-core==1.0.0, but you'll have azure-core 1.7.0 which is incompatible.
ERROR: helix-scripts 20191204.1 has requirement azure-storage-blob==12.0.0, but you'll have azure-storage-blob 12.5.0 which is incompatible.
ERROR: helix-scripts 20191204.1 has requirement certifi==2018.11.29, but you'll have certifi 2020.6.20 which is incompatible.
```

And the other one with this:
```
D:\workspace\_work\1\s\.packages\microsoft.dotnet.sharedframework.sdk\6.0.0-beta.21105.5\targets\sharedfx.targets(273,5): error : The following files are missing entries in the templated manifest: [D:\workspace\_work\1\s\src\installer\pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Runtime.sfxproj]
D:\workspace\_work\1\s\.packages\microsoft.dotnet.sharedframework.sdk\6.0.0-beta.21105.5\targets\sharedfx.targets(273,5): error : pgort140.dll. Add these file names with extensions to the 'PlatformManifestFileEntry' item group for the runtime pack and corresponding ref pack to include them in the platform manifest. [D:\workspace\_work\1\s\src\installer\pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Runtime.sfxproj]
##[error].packages\microsoft.dotnet.sharedframework.sdk\6.0.0-beta.21105.5\targets\sharedfx.targets(273,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) The following files are missing entries in the templated manifest:
pgort140.dll. Add these file names with extensions to the 'PlatformManifestFileEntry' item group for the runtime pack and corresponding ref pack to include them in the platform manifest.

Build FAILED.
```

**Pull request merged**
